src/main/kotlin/edu/carroll/polyominoes/service/account/RegisterServiceImpl.kt

```diff
@@ -0,0 +1,120 @@
+package edu.carroll.polyominoes.service.account
+
+import edu.carroll.polyominoes.jpa.model.Account
+import edu.carroll.polyominoes.jpa.repo.AccountRepository
+import org.slf4j.LoggerFactory
+import org.springframework.security.crypto.bcrypt.BCrypt
+import org.springframework.stereotype.Service
+
+@Service
+class RegisterServiceImpl(private val accountRepo: AccountRepository) : RegisterService {
+
+    companion object {
+        private val log = LoggerFactory.getLogger(RegisterServiceImpl::class.java)
+    }
+
+    /**
+     * Given a username, determine if the user's given username already exist in the database.
+     *
+     * @param username - An username representing a user's account
+     * @return true if the user's given username is in the database, false otherwise
+     */
+    override fun validateUsernameExist(username: String): Boolean {
+
+        if (username.isNullOrBlank()) {
+            log.warn("validateUsernameExist: username '{}' was null or blank", username)
+            return false
+        }
+
+        log.info("validateUsernameDoesNotExist: Checking if '{}' is linked to an account", username)
+        val users = accountRepo.findByUsernameIgnoreCase(username)
+
+        // We expect 0 or 1, so if we get more than 1, bail out as this is an error we don't deal with properly.
+        if (users.isNotEmpty()) {
+            if (users.size > 1) {
+                log.warn("validateUsernameExist: found {} users for '{}'", users.size, username)
+            } else {
+                log.debug("validateUsernameExist: found 1 user for '{}'", username)
+            }
+            return true
+        }
+
+        log.info("validateUsernameExist: '{}' is available", username)
+        return false
+    }
+
+    /**
+     * Given an email, determine if the user's given email already exist in the database.
+     *
+     * @param email - an email which represents the user account
+     * @return true if the user's given email is in the database, false otherwise
+     */
+    override fun validateEmailExist(email: String): Boolean {
+        log.info("validateUniqueEmail: Checking if '{}' is linked to an account", email)
+        val emails = accountRepo.findByEmailIgnoreCase(email)
```

Comment:
This fetch should be done AFTER the email field is checked for invalid data, as email could be invalid right now and we're fetching data from the DB with invalid information.